### PR TITLE
Persist the last-viewed pipeline in the project.

### PR DESCRIPTION
**Pull request by iannbing**
## Description

Currently, when user access a project, PipelineEditor always loads the first pipeline in the list.
This PR makes it so that Pipeline Editor will persist the UUID of the last-viewed pipeline in the project. If user enter the same project again, Pipeline Editor will load the last-viewed pipeline for user to continue their work.

This is achieved by saving a json object `{ [project_uuid]: pipeline_uuid }` in the localstorage.

This PR also fixes several bugs regarding routing:
- when moving forward/backward with the browser forward/backward button, Pipeline Editor doesn't load the pipeline.
- after creating a new pipeline, Pipeline Editor should open it immediately

## Checklist

- [x] The PR branch is set up to merge into `dev` instead of `master`.

### CODE REVIEW

**Comment by fruttasecca** [comment]
Not sure if related to changes in this PR but after creating a pipeine the list of sessions in the top left did not update, see
image

**Comment by fruttasecca** [comment]
Not sure if related to this PR or not, if some BE changes are involved let me know, see
https://www.loom.com/share/6ea43d4b9b774e7f82baec8f82c86c53

**Comment by fruttasecca** [comment]
>@fruttasecca How did you reproduce this?

Not really sure, it appeared to happen quite randomly, will try again later.

**Comment by fruttasecca** [comment]
@iannbing sorry for coming back late on this but I can indeed confirm that there might be something wrong, see video. Was on a rebuilt branch and the db was wiped, project just imported
https://www.loom.com/share/91bbc4462e734fac83772729eb024368

**Comment by iannbing** [comment]
> sorry for coming back late on this but I can indeed confirm that there might be something wrong, see video. Was on a rebuilt branch and the db was wiped, project just imported. I haven't tested this on dev.

@fruttasecca Thanks for helping test. I found that some issues were caused by SWR, and I had to change a lot (including the jest tests). Now I think all the reproducible issues are addressed.

@ricklamers  Browser forward/backward should work now.

I also encountered some weird issues though while testing, but couldn't reproduce them consistently. It seems that if I create enough amount of pipelines and did many operations by, for example, editing pipelines very fast, and go back and forth between pipelines using the browser forward/backward button, BE starts to throw 500 (similar to [the issue](https://github.com/orchest/orchest/pull/949#issuecomment-1128792737) that @fruttasecca encountered, but from dev tool, I saw 500 errors on GET requests to `/async/pipelines/json/<project_uuid>/<pipeline_uuid>` but I haven't look into BE logs.  Maybe something @fruttasecca can continue monitoring.

**Comment by fruttasecca** [comment]
> I saw 500 errors on GET requests to /async/pipelines/json/<project_uuid>/<pipeline_uuid> but I haven't look into BE logs. Maybe something @fruttasecca can continue monitoring.

I can take a look at this but it's likely that I won't be able to do it today, would this be blocking the PR?

**Comment by iannbing** [comment]
> I can take a look at this but it's likely that I won't be able to do it today, would this be blocking the PR?

@fruttasecca  No, it's not blocking this PR. I couldn't find the cause of this issue on the FE side, so just to raise awareness and also am curious what the root cause is.

**Comment by ricklamers** [comment]
This LGTM! Nice job @iannbing. Feel free to merge.

**Review by BrunoOrchest — APPROVED** [review]

File: services/orchest-webserver/client/src/components/DataTable.tsx [hunk 1]
```diff
@@ -197,15 +197,15 @@ function EnhancedTableHead<T>(props: EnhancedTableProps<T>) {
...
+  isLoading: boolean | undefined;
   sx?: SxProps<Theme>;
   skeletonSx?: SxProps<Theme>;
   onAuxClick?: (e: React.MouseEvent) => void;
 }> = ({ isLoading, sx, skeletonSx, onAuxClick, children }) => {
   const auxClickHandler = (e: React.MouseEvent) => {
     e.preventDefault();
     e.stopPropagation();
-    onAuxClick(e);
+    if (onAuxClick) onAuxClick(e);
```
[the remark this answers is not in the document]
Comment (iannbing) [reply services/orchest-webserver/client/src/components/DataTable.tsx:208]: Actually I'm surprised to learn that Typescript accepts this.

```
someEnterty.someProperty
someEnterty?.someProperty // an extra "?" 👍

something(args)
something?.(args) // this extra "." bothers me a bit  😕

```

From my understanding in Javascript, only an object is followed by a dot, and it almost becomes a reflex to me. 

`doSomething?.(args)` makes me think: `doSomething` could be an object and I might have made a naming mistake here, where in fact, `doSomething` is a function and the code is good 🤯

File: services/orchest-webserver/client/src/contexts/__tests__/useAutoFetchPipelines.test.tsx [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+    project1Pipelines,
+    project2Pipelines,
+  };
+};
+
+describe("useAutoFetchPipelines", () => {
+  const { result, waitForNextUpdate, rerender, unmount } = renderHook<
+    {
+      children?: null;
+      projectUuid?: string | undefined;
```
[the remark this answers is not in the document]
Comment (iannbing) [reply services/orchest-webserver/client/src/contexts/__tests__/useAutoFetchPipelines.test.tsx:69]: You're right. I wasn't very strict about the wrapper component for testing. Will change.

File: services/orchest-webserver/client/src/components/DataTable.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (BrunoOrchest) [reply services/orchest-webserver/client/src/components/DataTable.tsx:208]: Basically the `?` works for any type, but I'm ok with using the `if (do) do ()`, just used to use optional chaining on my code so in the future feel free to point it out :).

[the remark this answers is not in the document]
Comment (iannbing) [reply services/orchest-webserver/client/src/components/DataTable.tsx:208]: Hmm, seems that it's already like this [in Javascript](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining). I guess I need to get used to it 😅. Will change.

**Pull request merged**